### DISCUSSION TITLE
What happens with overprovisioning when there's no more space?

### DISCUSSION THREAD

**Discussion by refi64** [stratis-storage/project#463] [Q&A] [answered]
This is something I've been wondering about for a while. If my file systems are overprovisioned, but the pool runs out of physical space...what exactly happens? I've heard LVM's handling here is a bit messy, but I'm not sure what Stratis ends up doing.

There seems to be an [existing open issue for this](https://github.com/stratis-storage/project/issues/284), but the [now-implemented thin pool rework](https://github.com/stratis-storage/stratisd/issues/2814) mentions changes to out-of-space handling too, so I'm not sure if the behavior is different now.

**Comment by mulkieran** [comment, marked as answer]
@refi64 Thanks for the question! We are at present working on a blog post about the thinpool rework that we have done; it should be up within a few days on our website. I hope that will answer your questions.

**Reply by mulkieran** [reply]
It is up: https://stratis-storage.github.io/thin-provisioning-redesign/

**Reply by refi64** [reply]
Thanks! This certainly sounds really interesting, but...I'm not not sure what exactly happens. It says:

> When overprovisioning a storage stack, the stack can get into a **bad state** when the pool becomes full due to the filesystem being far larger than the pool backing it. 
> [...]
> For users that use Stratis for critical applications or the root filesystem, this mode prevents **certain failure cases that can be challenging to recover from**.

It's just not clear what this "bad state" is that's "challenging to recover from". e.g. does the underlying filesystem just get corrupted such that it's unreadable, or can you read but writes are broken / dropped (and fixing that is hard)?

**Reply by jbaublitz** [reply]
@bmr-cymru Feel free to chime in with any additional information, but here are my thoughts on the most likely cases.

This mainly boils down to some kernel-level communication issues between the filesystem and the block layer. The likely outcome you would see would be something like this:
* Pool gets filled up
* The system crashes and the XFS filesystem needs to do a journal replay
* The pool now has no space to allocate for the journal replay
* Your filesystem can no longer be mounted

There are numerous cases like this that we could outline but the overall idea is that the filesystem is expecting space to be available in lower layers. When that space is not available in the thin pool, the filesystem can get into states where it can no longer perform the necessary recovery given that that recovery requires space that the thin pool doesn't have.

These situations can sometimes be resolved by adding more storage to the thin pool, but sometimes the recovery is more involved. If you're interested in some of those cases, I think @bmr-cymru is the best person to talk with about that.

We would love to get to a state where XFS and dm-thin are on the same page about how much space is available for use, but that is going to require upstream work. We are actively talking about it, but the best solution isn't entirely clear yet. Our no overprovisioning mode is the best we can do without upstream work to ensure safety for users. If you want to take advantage of the overprovisioning benefits like dm-thin snapshot storage efficiency, we strongly recommend that you pay attention to the D-Bus signal indicating that all available space has been allocated so that you add additional storage space before you get into this position. That is a recent addition in this version intended to provide increased safety not just in the no-overprovisioning case but also in the overprovisioning case.

**Reply by refi64** [reply]
Ah that does make sense, matches with what I thought. Out of curiosity, would `norecovery` work to mount it in that case, since it wouldn't try to replay the journal?

**Reply by jbaublitz** [reply]
@refi64 That may work in that particular case, but I think the larger thing to keep in mind is the non-trivial recovery cases you can get yourself into when filling a pool. I'd actually be interested in your feedback here. We send out a D-Bus signal when storage has been fully allocated but not necessarily used. Would it be more useful to also send out a signal when the pool is about to run out of space to signal to stop writing or does that seem unlikely to be useful?

**Reply by refi64** [reply]
> Would it be more useful to also send out a signal when the pool is about to run out of space to signal to stop writing or does that seem unlikely to be useful?

TBH that would be very useful! Off the top of my head, I would probably wire it up to just remount the rootfs as ro if possible, which certainly isn't particularly *clean*, but it's much easier to work with localized application data corruption than actual filesystem issues. (Off the top of my head, I'm admittedly also not sure what the prospect of that actually working is!)

**Reply by jbaublitz** [reply]
That is unfortunately going to be very hard to do as you can only unmount if you kill all of the daemons running on the system that are accessing the root filesystem. We had considered that as a safety feature, but it proved to not be feasible. This is largely why we put the no overprovisioning mode in place (for the root filesystem case). Regardless, we will be providing a new property soon that will send out a signal when it's changed per your request to notify you when you're at a critical level of free space!

**Reply by jbaublitz** [reply]
@refi64 I wanted to follow up on the signal indicating critical levels of free space.

First I wanted to outline what we currently do. Our general approach is to allocate blocks of data in 50 GiB chunks. Whenever the low water mark is reached, 50 GiB of data space will be allocated or if there is less than that available, as much as possible will be allocated to the data device. Our low water mark is 15 GiB of free space remaining. If there is no additional space to allocate and the low water mark is reached (15 GiB left), we set `NoAllocSpace` to true and send out a signal indicating that no more room is available to allocate to the data device.

My question for you is whether 15 GiB is a reasonable amount of space left for a critical space signal if there is no storage left to allocate. If so, then we already support this feature under the `NoAllocSpace` property change signal. If you would like a critical signal at a lower amount of free space, then we should discuss the properties of what you're expecting (as in how much space would you like to see left before we send the signal). Do you have any thoughts here?

**Reply by refi64** [reply]
15 GiB should work for me as-is then!

Interestingly enough, by complete coincidence, I *did* manage to exhaust disk space, and it created a giant mess of...nothing, actually. Attempting to write new files have ENOSPC, but I could still delete some old files to free up space, after which everything went back to normal. Does the worse behavior mostly happen if you have existing snapshots left, in which case deleting files wouldn't actually free up any space, or was I just really lucky?

**Reply by jbaublitz** [reply]
You would probably see the worst kinds of cases if the thin pool is full and an unclean unmount occurs. A lot of the scenarios that I've heard are very difficult to recover from are log replays with a full thin pool.
